### PR TITLE
ADD test: Should be able to reroute to a path with not enough bandwidth available in case ignoreBandwidth=true

### PR DESCRIPTION
**Pull request by andriidovhan**
add test
"Should be able to reroute to a path with not enough bandwidth available in case ignoreBandwidth: true"

Close: #2063

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 1]
```diff
@@ -120,6 +120,60 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+
+        when: "Make the current path less preferable than alternatives"
+        def alternativePaths = allPaths.findAll { it != currentPath }
+        alternativePaths.each { pathHelper.makePathMorePreferable(it, currentPath) }
+
+        and: "Make all alternative paths to have not enough bandwidth to handle the flow"
+        def currentIsls = pathHelper.getInvolvedIsls(currentPath)
+        def changedIsls = alternativePaths.collect { altPath ->
+            def thinIsl = pathHelper.getInvolvedIsls(altPath).find {
+                !currentIsls.contains(it) && !currentIsls.contains(islUtils.reverseIsl(it))
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: You should rebase on develop. `islUtils.reverseIsl()` has been removed recently. Use `it.reversed` instead.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 2]
```diff
@@ -120,6 +120,60 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+            }
+            def newBw = flow.maximumBandwidth - 1
+            database.updateLinkMaxBandwidth(thinIsl, newBw)
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: `[thinIsl, thinIsl.reversed].each { database.updateLinkMaxBandwidth(it, newBw) }`

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 3]
```diff
@@ -120,6 +120,60 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+            database.updateLinkMaxBandwidth(islUtils.reverseIsl(thinIsl), newBw)
+            database.updateLinkAvailableBandwidth(thinIsl, newBw)
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: ditto.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 4]
```diff
@@ -120,6 +120,60 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        rerouteResponse.path.path != currentPath
+        rerouteResponse.path.path.each { assert it.seqId == seqId++ }
+
+        PathHelper.convert(northbound.getFlowPath(flow.id)) != currentPath
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.UP }
+
+        and: "Remove the flow, restore the bandwidth on ISLs, reset costs"
+        flowHelper.deleteFlow(flow.id)
+        changedIsls.each {
+            database.revertIslBandwidth(it)
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: ditto.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 5]
```diff
@@ -120,6 +120,60 @@ class IntentionalRerouteSpec extends BaseSpecification {
         [thinIsl, islUtils.reverseIsl(thinIsl)].each { database.revertIslBandwidth(it) }
     }
 
+    def "Should be able to reroute to a path with not enough bandwidth available in case ignoreBandwidth: true"() {
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Not a big deal, but in other tests we use style `ignore_bandwidth=true`.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 1 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Done.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 2 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Done.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 6]
```diff
@@ -184,6 +184,60 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        then: "The flow is rerouted because ignoreBandwidth:true"
+        int seqId = 0
+
+        rerouteResponse.rerouted
+        rerouteResponse.path.path != currentPath
+        rerouteResponse.path.path.each { assert it.seqId == seqId++ }
+
+        PathHelper.convert(northbound.getFlowPath(flow.id)) != currentPath
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.UP }
+
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: It makes sense to verify that all isls on the new path have their available bandwidth unchanged (the rerouted flow didn't change their available bandwidth)

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Added.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 7]
```diff
@@ -184,6 +184,64 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        rerouteResponse.rerouted
+        rerouteResponse.path.path != currentPath
+        rerouteResponse.path.path.each { assert it.seqId == seqId++ }
+
+        PathHelper.convert(northbound.getFlowPath(flow.id)) != currentPath
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.UP }
+        def cPath = PathHelper.convert(northbound.getFlowPath(flow.id))
+        def allIsl = northbound.getAllLinks()
+        def firstIsl = pathHelper.getInvolvedIsls(cPath).collect { islUtils.getIslInfo(allIsl, it).get() }[0]
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Remove L233 and do
`def firstIsl = islUtils.getIslInfo(pathHelper.getInvolvedIsls(cPath)[0]).get()`
this will not cycle through all the isls because you don't actually need it in current implementation and just use the first one

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 8]
```diff
@@ -184,6 +184,64 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        firstIsl.availableBandwidth < flow.maximumBandwidth
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Don't understand what is being verified here. You should: take isl bw before reroute, and verify that it is the same after reroute. The `firstIsl.availableBandwidth < flow.maximumBandwidth` basically does nothing. The `firstIsl.availableBandwidth` may be 0 or even drop into negative value and your test will pass. This is an autopass verification.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 9]
```diff
@@ -184,6 +184,64 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+
+        then: "The flow is rerouted because ignoreBandwidth:true"
+        int seqId = 0
+
+        rerouteResponse.rerouted
+        rerouteResponse.path.path != currentPath
+        rerouteResponse.path.path.each { assert it.seqId == seqId++ }
+
+        PathHelper.convert(northbound.getFlowPath(flow.id)) != currentPath
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.UP }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: I think it worth to add another `and: "Something other happened"` block here to describe what is additionally being verified except for 'The flow is rerouted because ignoreBandwidth:true'

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 10]
```diff
@@ -184,6 +184,64 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        def cPath = PathHelper.convert(northbound.getFlowPath(flow.id))
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: You did this operation at L230 already. Let's have 'updatedPath' variable instead of two calls.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 11]
```diff
@@ -184,6 +184,64 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        def allIsl = northbound.getAllLinks()
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: allIsls or allLinks

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 8 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: I will compare it against the newBw variable
just to avoid introducing new variable
it should be accepted here, cause we just want to be sure that available bandwidth was not changed after rerouting

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 9 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: added: "Updated bandwidth was not changed by system"

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 10 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: done

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 11 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: allIsls
fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 12]
```diff
@@ -187,6 +187,65 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+
+        rerouteResponse.rerouted
+        rerouteResponse.path.path != currentPath
+        rerouteResponse.path.path.each { assert it.seqId == seqId++ }
+
+        and: "Available bandwidth was not changed while rerouting due to ignoreBandwidth=true"
+        def updatedPath = PathHelper.convert(northbound.getFlowPath(flow.id))
+        updatedPath != currentPath
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.UP }
+        def allLinks = northbound.getAllLinks()
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Sorry, didn't notice it before. We can drop this line at all.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 13]
```diff
@@ -187,6 +187,65 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        def firstIsl = islUtils.getIslInfo(allLinks, pathHelper.getInvolvedIsls(updatedPath)[0]).get()
+        firstIsl.availableBandwidth == newBw
+
+        and: "Remove the flow, restore the bandwidth on ISLs, reset costs"
+        flowHelper.deleteFlow(flow.id)
+        changedIsls.each {
+            database.revertIslBandwidth(it)
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: revert -> reset

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 9 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: changed to: "Available bandwidth was not changed while rerouting due to ignoreBandwidth=true"

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 14]
```diff
@@ -187,6 +187,64 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        int seqId = 0
+
+        rerouteResponse.rerouted
+        rerouteResponse.path.path != currentPath
+        rerouteResponse.path.path.each { assert it.seqId == seqId++ }
+
+        and: "Available bandwidth was not changed while rerouting due to ignoreBandwidth=true"
+        def updatedPath = PathHelper.convert(northbound.getFlowPath(flow.id))
+        updatedPath != currentPath
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.UP }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: L234-L236 belong to `then: "The flow is rerouted"`, these verifications are not related to bandwidth at all

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 15]
```diff
@@ -187,6 +187,64 @@ class IntentionalRerouteSpec extends BaseSpecification {
...
+        def firstIsl = islUtils.getIslInfo(pathHelper.getInvolvedIsls(updatedPath)[0]).get()
+        firstIsl.availableBandwidth == newBw
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: There is no guarantee that the first isl in `getInvolvedIsls` will have `newBw`. On L210-221 you updated a single ISL for each alterantive path, but there is no guarantee that it will always be the first one(for example when current and alt path have the same ISL as the first one). This depends on topology and on ISLs does the current path use. The guaranteed list of ISLs that have `newBw` is stored in the `changedIsls` variable. You can leverage this somehow. Otherwise - this is just a coincidence that this verification works

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 12 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 13 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 14 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: I've just tried to store the `updatedPath` variable more close.
Fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy (hunk 15 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy]: Fixed
